File: formal/runtime/java/dafny/DafnyMap.java
// Copyright by the contributors to the Dafny Project
// SPDX-License-Identifier: MIT
//
// EACL target-cost refinement. See DafnySet.java in this directory.

package dafny;

import clojure.lang.IPersistentMap;
import clojure.lang.PersistentHashMap;
import java.util.*;
import java.util.function.BiConsumer;

public class DafnyMap<K, V> {
    private IPersistentMap innerMap;

    public DafnyMap() {
        innerMap = PersistentHashMap.EMPTY;
    }

    public DafnyMap(Map<K, V> values) {
        assert values != null : "Precondition Violation";
        innerMap = PersistentHashMap.EMPTY;
        for (Map.Entry<K, V> entry : values.entrySet()) {
            innerMap = innerMap.assoc(entry.getKey(), entry.getValue());
        }
    }

    private DafnyMap(IPersistentMap persistentMap) {
        innerMap = persistentMap;
    }

    public static <K, V> DafnyMap<K, V> empty() {
        return new DafnyMap<>();
    }

    @SafeVarargs
    public static <K, V> DafnyMap<K, V> fromElements(
            Tuple2<K, V> ... pairs) {
        DafnyMap<K, V> result = new DafnyMap<>();
        for (Tuple2<K, V> pair : pairs) {
            result.innerMap =
                result.innerMap.assoc(pair.dtor__0(), pair.dtor__1());
        }
        return result;
    }

    @SuppressWarnings("unchecked")
    public static <K, V> TypeDescriptor<DafnyMap<? extends K, ? extends V>>
            _typeDescriptor(
                TypeDescriptor<K> keyType,
                TypeDescriptor<V> valueType) {
        return TypeDescriptor.referenceWithDefault(
                (Class<DafnyMap<? extends K, ? extends V>>)
                    (Class<?>) DafnyMap.class,
                DafnyMap.empty());
    }

    public boolean contains(Object key) {
        return innerMap.containsKey(key);
    }

    @SuppressWarnings("unchecked")
    public static <K, V> DafnyMap<K, V> update(
            DafnyMap<? extends K, ? extends V> source,
            K key,
            V value) {
        return new DafnyMap<>(
            ((IPersistentMap) source.innerMap).assoc(key, value));
    }

    @Override
    public boolean equals(Object value) {
        if (this == value) {
            return true;
        }
        if (value == null || getClass() != value.getClass()) {
            return false;
        }
        DafnyMap<?, ?> other = (DafnyMap<?, ?>) value;
        return innerMap.equals(other.innerMap);
    }

    @Override
    public int hashCode() {
        return innerMap.hashCode();
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder("map[");
        String separator = "";
        for (Map.Entry<K, V> entry : entries()) {
            result.append(separator)
                .append(Helpers.toString(entry.getKey()))
                .append(" := ")
                .append(Helpers.toString(entry.getValue()));
            separator = ", ";
        }
        return result.append("]").toString();
    }

    public void forEach(BiConsumer<? super K, ? super V> action) {
        for (Map.Entry<K, V> entry : entries()) {
            action.accept(entry.getKey(), entry.getValue());
        }
    }

    @SuppressWarnings("unchecked")
    public static <K, V> DafnyMap<? extends K, ? extends V> merge(
            DafnyMap<? extends K, ? extends V> left,
            DafnyMap<? extends K, ? extends V> right) {
        assert left != null : "Precondition Violation";
        assert right != null : "Precondition Violation";
        if (left.isEmpty()) {
            return right;
        }
        if (right.isEmpty()) {
            return left;
        }
        DafnyMap<K, V> result =
            new DafnyMap<>((IPersistentMap) right.innerMap);
        left.forEach((key, value) -> {
            if (!result.contains(key)) {
                result.innerMap = result.innerMap.assoc(key, value);
            }
        });
        return result;
    }

    @SuppressWarnings("unchecked")
    public static <K, V> DafnyMap<? extends K, ? extends V> subtract(
            DafnyMap<? extends K, ? extends V> source,
            DafnySet<? extends K> keys) {
        assert source != null : "Precondition Violation";
        assert keys != null : "Precondition Violation";
        if (source.isEmpty() || keys.isEmpty()) {
            return source;
        }
        IPersistentMap result = source.innerMap;
        for (K key : keys.Elements()) {
            result = result.without(key);
        }
        return new DafnyMap<K, V>(result);
    }

    public int size() {
        return innerMap.count();
    }

    public int cardinalityInt() {
        return size();
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    @SuppressWarnings("unchecked")
    public V get(Object key) {
        return (V) innerMap.valAt(key);
    }

    public DafnySet<K> keySet() {
        return new DafnySet<>(asJavaMap().keySet());
    }

    public DafnySet<V> valueSet() {
        return new DafnySet<>(asJavaMap().values());
    }

    @SuppressWarnings("unchecked")
    public <KK, VV> DafnySet<? extends Tuple2<KK, VV>> entrySet() {
        ArrayList<Tuple2<K, V>> values = new ArrayList<>();
        for (Map.Entry<K, V> entry : entries()) {
            values.add(
                new Tuple2<K, V>(entry.getKey(), entry.getValue()));
        }
        return (DafnySet<? extends Tuple2<KK, VV>>)
            (Object) new DafnySet<Tuple2<K, V>>(values);
    }

    @SuppressWarnings("unchecked")
    private Map<K, V> asJavaMap() {
        return (Map<K, V>) innerMap;
    }

    private Set<Map.Entry<K, V>> entries() {
        return asJavaMap().entrySet();
    }
}

File: formal/runtime/js/dafny-map.js
  // EACL target-cost refinement. See dafny-set.js in this directory.
  $module.Map = class Map {
    constructor(inner) {
      this._inner = inner === undefined ? Immutable.Map() : inner;
    }
    static get Default() {
      return Map.Empty;
    }
    static get Empty() {
      if (this._empty === undefined) {
        this._empty = new Map();
      }
      return this._empty;
    }
    static of(...entries) {
      const result = new Map();
      for (const [key, value] of entries) {
        result.updateUnsafe(key, value);
      }
      return result;
    }
    get length() {
      return this._inner.size;
    }
    slice() {
      return new Map(this._inner);
    }
    [Symbol.iterator]() {
      return this.entries();
    }
    *entries() {
      for (const pair of this._inner.valueSeq()) {
        yield pair;
      }
    }
    findIndex(key) {
      let index = 0;
      for (const [candidate] of this) {
        if (_dafny.areEqual(candidate, key)) {
          return index;
        }
        index++;
      }
      return this.length;
    }
    get(key) {
      const pair = this._inner.get(new DafnyKey(key));
      return pair === undefined ? undefined : pair[1];
    }
    contains(key) {
      return this._inner.has(new DafnyKey(key));
    }
    update(key, value) {
      return new Map(
        this._inner.set(new DafnyKey(key), [key, value]));
    }
    updateUnsafe(key, value) {
      this._inner =
        this._inner.set(new DafnyKey(key), [key, value]);
      return this;
    }
    equals(other) {
      if (this === other) {
        return true;
      }
      if (!(other instanceof Map) || this.length !== other.length) {
        return false;
      }
      for (const [key, value] of this) {
        if (!other.contains(key) ||
            !_dafny.areEqual(value, other.get(key))) {
          return false;
        }
      }
      return true;
    }
    toString() {
      return "map[" +
        Array.from(
          this,
          ([key, value]) =>
            _dafny.toString(key) + " := " + _dafny.toString(value)
        ).join(", ") +
        "]";
    }
    get Keys() {
      const result = new _dafny.Set();
      for (const [key] of this) {
        result.add(key);
      }
      return result;
    }
    get Values() {
      const result = new _dafny.Set();
      for (const [, value] of this) {
        result.add(value);
      }
      return result;
    }
    get Items() {
      const result = new _dafny.Set();
      for (const [key, value] of this) {
        result.add(_dafny.Tuple.of(key, value));
      }
      return result;
    }
    Merge(other) {
      let inner = other._inner;
      for (const [key, value] of this) {
        const wrapped = new DafnyKey(key);
        if (!inner.has(wrapped)) {
          inner = inner.set(wrapped, [key, value]);
        }
      }
      return new Map(inner);
    }
    Subtract(keys) {
      if (this.length === 0 || keys.length === 0) {
        return this;
      }
      let inner = this._inner;
      for (const key of keys) {
        inner = inner.remove(new DafnyKey(key));
      }
      return new Map(inner);
    }
  }
